crates/vsock/src/main.rs

```diff
@@ -10,38 +10,81 @@ mod vsock_conn;
 
 use std::{convert::TryFrom, sync::Arc};
 
-use clap::Parser;
+use crate::vhu_vsock::{Error, Result, VhostUserVsockBackend, VsockConfig};
+use clap::{Args, Parser};
 use log::{info, warn};
+use serde::Deserialize;
 use vhost::{vhost_user, vhost_user::Listener};
 use vhost_user_backend::VhostUserDaemon;
 use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
 
-use crate::vhu_vsock::{Error, Result, VhostUserVsockBackend, VsockConfig};
-
-#[derive(Parser, Debug)]
-#[clap(version, about, long_about = None)]
-struct VsockArgs {
+#[derive(Args, Debug, Deserialize)]
+struct VsockParam {
     /// Context identifier of the guest which uniquely identifies the device for its lifetime.
-    #[clap(long, default_value_t = 3)]
+    #[arg(long, default_value_t = 3, conflicts_with = "config")]
     guest_cid: u64,
 
     /// Unix socket to which a hypervisor connects to and sets up the control path with the device.
-    #[clap(long)]
+    #[arg(long, conflicts_with = "config")]
     socket: String,
 
     /// Unix socket to which a host-side application connects to.
-    #[clap(long)]
+    #[arg(long, conflicts_with = "config")]
     uds_path: String,
 }
 
+#[derive(Parser, Debug)]
+#[command(version, about, long_about = None)]
+struct VsockArgs {
+    #[command(flatten)]
+    param: Option<VsockParam>,
+
+    /// Load from a given configuration file
+    #[arg(long)]
+    config: Option<String>,
+}
+
+impl VsockArgs {
+    pub fn parse_config(&self) -> Option<VsockConfig> {
+        match &self.config {
+            Some(c) => {
+                match config::Config::builder()
+                    .add_source(config::File::new(c.as_str(), config::FileFormat::Yaml))
+                    .build()
+                {
+                    Ok(s) => {
+                        let mut v = s.get::<Vec<VsockParam>>("vms").unwrap();
+                        if v.len() != 1 {
+                            return None;
+                        }
+                        v.pop()
+                            .map(|vm| VsockConfig::new(vm.guest_cid, vm.socket, vm.uds_path))
+                    }
+                    Err(_) => None,
+                }
+            }
+            _ => None,
+        }
+    }
+}
+
 impl TryFrom<VsockArgs> for VsockConfig {
     type Error = Error;
 
     fn try_from(cmd_args: VsockArgs) -> Result<Self> {
-        let socket = cmd_args.socket.trim().to_string();
-        let uds_path = cmd_args.uds_path.trim().to_string();
-
-        Ok(VsockConfig::new(cmd_args.guest_cid, socket, uds_path))
+        // we try to use the configuration first, if failed,  then fall back to the manual settings.
+        let config: Option<Self> = cmd_args.parse_config();
+        if let Some(c) = config {
```

Comment:
While at this place I belive a match may look better, since we want to write code for all match cases.

Reply:
No, but something like this:

        match config {
            Some(c) => Ok(c),
	    _ => {
              cmd_args.param.map_or(Err(Error::ConfigParse), |p| {
                  Ok(Self::new(
                      p.guest_cid,
                      p.socket.trim().to_string(),
                      p.uds_path.trim().to_string(),
                  ))
              })
	    }
       }